src/gui/maptools/qgsmaptoolcapturelayergeometry.cpp

```diff
@@ -37,11 +37,9 @@ void QgsMapToolCaptureLayerGeometry::geometryCaptured( const QgsGeometry &geomet
       break;
     case QgsMapToolCapture::CaptureLine:
     case QgsMapToolCapture::CapturePolygon:
-      //does compoundcurve contain circular strings?
       //does provider support circular strings?
-      const bool hasCurvedSegments = captureCurve()->hasCurvedSegments();
```

Comment:
I don't understand this fix, does `hasCurvedSegments()` return `false` even when there are curves?

Reply:
`hasCurvedSegments()` is correct.

The problem here is the following: if the digitized feature does not have curved segments, we must still keep it a CompoundCurve if the layer/provider supports curves.

That way, we can still use the "convert vertex to curve" without the #53818 issue

Reply:
I am probably not correctly understanding the term "digitized feature" here. My understanding: if we convert a vertex to a curve point, the "digitized feature" now has curved segments and the check should return true. Or is the "digitized feature" something else than the geometry of the rubberband?

Reply:
You are right, when the user digitizes his feature using curves it works that way.

The problem is when the user digitizes his feature without curves, then switches to the vertex tool (so digitizing is finished), selects a vertex, and converts it to a curved vertex with the `O` keyboard shortcut.

Reply:
So the "digitized feature" does not contain the edited geometry including the curve that was created with `O`? It's just the "old" unmodified geometry?

Reply:
In the end (user digitizes the feature without curve, selects vertex tool, selects a vertex, hits `O`) the newly created curve is segmentized (see capture).

Reply:
I think the first part is not even needed (you can also take a pre-existing feature).
So let me ask differently: what does `captureCurve()` represent exactly? The geometry of the unmodified feature (before pressing `O`)?

Reply:
Yes, `captureCurve()` is the geometry before pressing `O`. It's the digitized geometry, after right-clicking.

Reply:
Do we have access to the geometry including the modification?

Reply:
In fact, if the user digitizes 3 features without curves (for example) in a layer compatible with curves, if those features are not saved in the layer yet, the bug will occur if we try to convert a vertex.

That's why I specify "digitized features" and not "features".

Because a geometry without curves, already available from a data source (already saved in the layer), has been forced by the "save to layer / commit" action to have a correct geometry type (CompoundCurve), and everything is fine when converting one vertex to curve vertex.

Difficult to explain, but we'll get to it :D

Reply:
Ok, got it. Thanks!

Do we have access to a modified geometry? I.e. if we have not added a curve, can we leave the feature uncurved (this seems to have been the intention of this)

Reply:
> Do we have access to a modified geometry?

Sorry, I don't understand the question...

> I.e. if we have not added a curve, can we leave the feature uncurved (this seems to have been the intention of this)

The digitizing gives `captureCurve()` as a result, which is a CompoundCurve.

If we have not added a curve, 
- before this PR the geometry was converted into a LineString (whatever the layer type is)
- after this PR the geometry won't be changed and remain CompoundCurve if the layer is compatible, or be converted into a LineString if the layer is not compatible

Reply:
Hmmm... A check could be done on `g` (that's what I meant), but I'm not actually sure if that brings value.